cupy/core/core.pyx

```diff
@@ -2363,6 +2378,14 @@ cpdef ndarray _convert_object_with_cuda_array_interface(a):
         dev_id = device.get_device_id()
     mem = memory_module.UnownedMemory(ptr, nbytes, a, dev_id)
     memptr = memory.MemoryPointer(mem, 0)
+    # the v3 protocol requires an immediate synchronization, unless
+    # 1. the stream is not set (ex: from v0 ~ v2) or is None
+    # 2. users explicitly overwrite this requirement
+    stream_ptr = desc.get('stream')
+    if stream_ptr is not None:
```

Comment:
Do we need to consider the case of `stream1` = 1 or 2?

Reply:
Strictly speaking, no. We should sync any given stream unless this behavior is explicitly turned off by users. The issue here is at this point we as Consumer could be on a different stream, so we must ensure the data is ready for subsequent operations. From this point of view, 1 or 2 are no different from other streams. 

In previous commits I tried to compare the current stream (in CuPy) with the external stream (from CAI), and decide whether to sync or not, but I felt it's error-prone so I removed it.

I think the CAI v3 protocol gave a [very complex example](https://numba.readthedocs.io/en/latest/cuda/cuda_array_interface.html#interface-requirements) (need to scroll down a bit) that could ruin any attempt of sync policy. For example, a foreign library (Producer) could operate the data on multiple streams, and export a totally different one which internally waits on all aforementioned streams. Then the Consumer must sync if correctness is the priority (or if the Producer's internal is unknown to Users and thus impossible to control).

A potential micro-optimization for the external stream being 1 (`cudaStreamLegacy`) is to rely on its implicit synchronization and don't sync. But again, if we're on a non-blocking stream we're doomed. Am I making sense?

Reply:
I think it's OK to synchronize on default and let users decide that synchronization as the current implementation does. :+1:

How about on `cuda.stream.ExternalStream(stream_ptr)` for the case of `stream` = 1 or 2?

Reply:
Thanks, @asi1024!

> How about on `cuda.stream.ExternalStream(stream_ptr)` for the case of `stream` = 1 or 2?

Could you clarify on this? Isn't this already done?

Reply:
Sorry for not explaining enough. In my understanding, More additional handling is needed for these cases, and you are planning to make another PR for complete support. Is it correct?
If so, could you fix to raise `NotImplementedError` in this PR, and restore in the other PR?

Reply:
No, actually this is enough. The point here is whatever pointer here, if given, will be synchronized.

Reply:
Sorry, I was misunderstanding something.